pkg/plugins/openai/cmd/main/main.go

```diff
@@ -0,0 +1,345 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"os"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+
+	"golang.org/x/time/rate"
+	"google.golang.org/protobuf/types/known/timestamppb"
+
+	"github.com/google/uuid"
+	"github.com/hashicorp/go-plugin"
+	commonconfig "github.com/opencost/opencost-plugins/pkg/common/config"
+	openaiplugin "github.com/opencost/opencost-plugins/pkg/plugins/openai/openaiplugin"
+	"github.com/opencost/opencost/core/pkg/log"
+	"github.com/opencost/opencost/core/pkg/model/pb"
+	"github.com/opencost/opencost/core/pkg/opencost"
+	ocplugin "github.com/opencost/opencost/core/pkg/plugin"
+	"github.com/opencost/opencost/core/pkg/util/timeutil"
+)
+
+// handshakeConfigs are used to just do a basic handshake between
+// a plugin and host. If the handshake fails, a user friendly error is shown.
+// This prevents users from executing bad plugins or executing a plugin
+// directory. It is a UX feature, not a security feature.
+var handshakeConfig = plugin.HandshakeConfig{
+	ProtocolVersion:  1,
+	MagicCookieKey:   "PLUGIN_NAME",
+	MagicCookieValue: "openai",
+}
+
+const openAIUsageURLFmt = "https://api.openai.com/v1/usage?date=%s"
+const openAIBillingURLFmt = "https://api.openai.com/v1/dashboard/billing/usage/export?exclude_project_costs=false&file_format=json&new_endpoint=true&project_id&start_date=%s&end_date=%s"
+const openAIAPIDateFormat = "2006-01-02"
+
+// Implementation of CustomCostSource
+type OpenAICostSource struct {
+	rateLimiter *rate.Limiter
+	config      *openaiplugin.OpenAIConfig
+}
+
+func (d *OpenAICostSource) GetCustomCosts(req *pb.CustomCostRequest) []*pb.CustomCostResponse {
+	results := []*pb.CustomCostResponse{}
+
+	targets, err := opencost.GetWindows(req.Start.AsTime(), req.End.AsTime(), req.Resolution.AsDuration())
+	if err != nil {
+		log.Errorf("error getting windows: %v", err)
+		errResp := pb.CustomCostResponse{
+			Errors: []string{fmt.Sprintf("error getting windows: %v", err)},
+		}
+		results = append(results, &errResp)
+		return results
+	}
+
+	if req.Resolution.AsDuration() != timeutil.Day {
+		log.Infof("openai plugin only supports hourly resolution")
```

Comment:
Should this say `only supports daily resolution`?

Reply:
And is this the proper log level?

Reply:
yup good catch